SCC-3760 - EDD page UI

dgcohen opened the pull request:
## Ticket:

- JIRA ticket [SCC-3760](https://newyorkpubliclibrary.atlassian.net/browse/SCC-3760)

## This PR does the following:

- Adds the EDD request page and associated components (EDD Form)
- Adds validation logic to focus on first invalid field on submission
- Extracts hold request banner into its own component and adds logic for different statuses in EDD form (still haven't implemented statuses in Hold page since there are only 2 states)
- Adds basic tests (more detailed tests forthcoming)

## How has this been tested?

- Added basic unit tests

## Accessibility concerns or updates

NA

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I updated the CHANGELOG with the appropriate information and JIRA ticket number (if applicable).
- [x] I have added relevant accessibility documentation for this pull request.
- [x] All new and existing tests passed.


[SCC-3760]: https://newyorkpubliclibrary.atlassian.net/browse/SCC-3760?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

charmingduchess reviewed (changes requested):
Looks good!
I think it would be a good idea to rename `holdUtils` and `holdTypes` to `holdPageUtils` and `holdPageTypes`

charmingduchess commented on src/reducers/eddFormReducer.ts:
```diff
@@ -0,0 +1,17 @@
+import type { EDDRequestParams, EDDFormAction } from "../types/holdTypes"
+
+export const eddFormReducer = (
```
@dgcohen is this possible? this seems redundant to the other reducer.

charmingduchess replied:
@dgcohen what do you think about this?

EdwinGuzman commented on src/components/HoldPages/EDDRequestForm.tsx:
```diff
@@ -0,0 +1,282 @@
+    }, {})
+  )
+
+  // Separate validation from input change to validate on tabbing (blur) and to prevent validation on every keystroke on fields other than email
```
👍

EdwinGuzman commented on src/components/HoldPages/EDDRequestForm.tsx:
```diff
@@ -0,0 +1,282 @@
+      action={`${BASE_URL}/api/hold/request/${holdId}/edd`}
+      onSubmit={validateAndSubmit}
+      mb="l"
+      maxWidth={{ lg: "75%" }}
```
Is this from the figma design?

EdwinGuzman commented on src/components/HoldPages/HoldRequestBanner.tsx:
```diff
@@ -45,37 +47,42 @@ const HoldRequestBanner = ({
+                <Button
+                  id="hold-contact"
+                  onClick={() =>
+                    onContact({
```
👍

EdwinGuzman commented on src/utils/holdPageUtils.ts:
```diff
@@ -0,0 +1,85 @@
+        case "email":
+          return {
+            key: "email",
+            isInvalid: !fieldValue.length || !EMAIL_REGEX.test(fieldValue),
```
I forgot already, but is this similar or the same validation as the My Account email test?

charmingduchess replied:
I don't know that Emma's work is referencing this constant yet, but this constant is based on Emma's validation regex.

EdwinGuzman reviewed (commented):
Looks great but have a few questions.

One big question about how the form works, when it fails I get the error message and that's fine, but all my data is gone. I would expect it to stay on the page for me to edit and make any updates but the form clears. We can discuss this as a group but I think the existing form does this, and in general, I think we should maintain the user data in forms when they fail.

charmingduchess commented:
> One big question about how the form works, when it fails I get the error message and that's fine, but all my data is gone. I would expect it to stay on the page for me to edit and make any updates but the form clears. We can discuss this as a group but I think the existing form does this, and in general, I think we should maintain the user data in forms when they fail.
I feel like I remember this persisting user data being implemented when I first started working here, and we probably should include it again.

charmingduchess commented on src/components/HoldPages/EDDRequestForm.tsx:
```diff
@@ -0,0 +1,282 @@
+    e.preventDefault()
+    const target = e.target as HTMLInputElement
+
+    dispatch({
```
I think this reducer is overkill for the single kind of input we're handling here.
I'd love to see it removed and instead just using
`const [eddFormState, setEddFormstate] = useState(initialFormState)`
and then calling `setEddFormState` instead of a dispatch to the reducer.

dgcohen replied:
Good call! replace reducer with useState

dgcohen commented on __test__/pages/hold/eddRequestPage.test.tsx:
```diff
@@ -0,0 +1,176 @@
+      ])
+    })
+  })
+  describe("aeon redirect handling", () => {
```
http://local.nypl.org:8080/research/research-catalog/hold/request/b21086080-i34439033/edd

dgcohen commented:
> Looks great but have a few questions.
> 
> One big question about how the form works, when it fails I get the error message and that's fine, but all my data is gone. I would expect it to stay on the page for me to edit and make any updates but the form clears. We can discuss this as a group but I think the existing form does this, and in general, I think we should maintain the user data in forms when they fail.

Fixed this by moving the state up to the page component level. I also removed the reducer and replaced with useState, but that was more to simplify things and didn't contribute to the fix.

EdwinGuzman commented on src/config/constants.ts:
```diff
@@ -139,3 +142,55 @@ export const NYPL_LOCATIONS = {
 export const HOLD_PAGE_HEADING = "Request for on-site use"
 export const EDD_PAGE_HEADING = "Request scan"
+
+export const EDD_FORM_FIELD_COPY = {
```
Good idea.

EdwinGuzman reviewed (approved):
Let's keeping move forward!
